x/slashing/src/keeper.rs

```diff
@@ -87,6 +103,211 @@ impl<SK: StoreKey, PSK: ParamsSubspaceKey, SSK: SlashingStakingKeeper<SK, M>, M:
+                self.set_validator_missed_block_bit_array(ctx, &cons_addr, index, false);
+                sign_info.missed_blocks_counter -= 1;
+            }
+            _ => {
+                // Array value at this index has not changed, no need to update counter
+            }
+        }
+
+        let min_signed_per_window = params.min_signed_per_window_u32();
+
+        if !signed {
+            ctx.append_events(vec![Event {
+                r#type: "liveness".to_string(),
+                attributes: vec![
+                    EventAttribute {
+                        key: "address".into(),
+                        value: cons_addr.clone().to_string().into(),
+                        index: false,
+                    },
+                    EventAttribute {
+                        key: "missed_blocks".into(),
+                        value: sign_info.missed_blocks_counter.to_string().into(),
+                        index: false,
+                    },
+                    EventAttribute {
+                        key: "height".into(),
+                        value: height.to_string().into(),
+                        index: false,
+                    },
+                ],
+            }]);
+
+            // TODO: how do we log?
+            tracing::debug!(
+                name: "absent validator",
+                // TODO: what is target?
+                target: "module::slashing",
+                ?height,
+                validator = cons_addr.to_string(),
+                missed = sign_info.missed_blocks_counter,
+                treshold = min_signed_per_window,
+            );
+            // logger.Debug(
+            //     "absent validator",
+            //     "height", height,
+            //     "validator", consAddr.String(),
+            //     "missed", signInfo.MissedBlocksCounter,
+            //     "threshold", minSignedPerWindow,
+            // )
+        }
+
+        let min_height = sign_info.start_height + params.signed_blocks_window as u32;
+        let max_missed = params.signed_blocks_window as u32 - min_signed_per_window;
+
+        // if we are past the minimum height and the validator has missed too many blocks, punish them
+        if height > min_height && sign_info.missed_blocks_counter > max_missed {
+            let validator_is_jailed = self
+                .staking_keeper
+                .validator_by_cons_addr(ctx, &cons_addr)
+                .unwrap_gas()
+                .map(|v| v.is_jailed())
+                .unwrap_or_default();
+
+            if validator_is_jailed {
+                // Downtime confirmed: slash and jail the validator
+                // We need to retrieve the stake distribution which signed the block, so we subtract ValidatorUpdateDelay from the evidence height,
+                // and subtract an additional 1 since this is the LastCommit.
+                // Note that this *can* result in a negative "distributionHeight" up to -ValidatorUpdateDelay-1,
```

Comment:
Our `distribution_height` is a `u32` so this isn't true.

Reply:
I'll fix

Reply:
done